ByteString optimisations of methods in HTTP parsing hot-path

ktoso opened the pull request:
Shaves of a lot of abstraction (mostly by avoiding the `MultiByteArrayIterator` as much as possible (we've seen it show up significantly in FlameGraphs of HTTP stress tests.

![flamegraph-6644-mexed-cpus_svg](https://cloud.githubusercontent.com/assets/120979/16954392/057ae912-4dd1-11e6-89b9-d4f0e5ed7204.png)

This PR speeds up various (the ones that are called _all the time_ in the HTTP parsers) ByteString operations in the range of `~10x`. 
Yes, it does sacrifice one-abstraction (iterator) to rule them all in the process, 10x is likely worth that though. I will measure overall impact on HTTP throughput now.

Focus was `drop`, `slice`. `take` and `dropRight` came well together with this work, so optimised them as well.
Increased test coverage of ByteString (some methods were not scala-check-ed, they now are), and also added explicit tests for cases that I found tricky (some found by scalacheck). Profiled impls using JFR, via `jmh:run -prof jmh.extras.JFR`, difference in allocation rates (memory pressure) and hot methods very visible:

![oracle_java_mission_control](https://cloud.githubusercontent.com/assets/120979/16954318/bc919cfa-4dd0-11e6-9078-0eb8f8daa1e0.png)

Hot methods list also shows that we're doing far less "dancing around the job":
![oracle_java_mission_control](https://cloud.githubusercontent.com/assets/120979/16954330/c7e36f2a-4dd0-11e6-8f2a-ee1d1c60f3c0.png)

Resolves #20992 
Resolves #15965

ktoso committed:
=act #20992 prepare benchmarks for ByteString optimisations

ktoso committed:
=act #20992 optimise common ByteString operations: drop,take,slice...

ktoso committed:
=act,htc #15965 add ByteString.decodeString(java.nio.charsets.Charset)

ktoso commented:
Benchmark results are the appropriate benchmark files.

johanandren commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -124,29 +125,47 @@ object ByteString {
 
     override def length: Int = bytes.length
 
+    @deprecated("Avoid `iterator` in performance sensitive code, call ops directly on ByteString instead", "2.4.9")
```
Should it really be deprecated, not just scaladoc:ed warning?

ktoso replied:
Open question, I guess. It helped me figure out where people call it.
I don't mind keeping the warning as `//` comment

viktorklang replied:
+1 on not deprecating it

ktoso replied:
roger that then 👍

johanandren commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -204,6 +230,35 @@ object ByteString {
+      dropRight1(n)
+
+    /** INTERNAL API */
+    private[akka] def dropRight1(n: Int): ByteString1 =
```
why the separate private method?

ktoso replied:
to not hide the fact that it's a `ByteString1` which is statically needed by `ByteStrings` (it contains `Vector[ByteString1]`. So instead of casting blindly, we can use this safe method

ktoso replied:
Users should never know about the existence of 1 / 1C etc

johanandren replied:
Lets not tell them then.

johanandren commented:
Cool. LGTM!

viktorklang commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -124,29 +125,47 @@ object ByteString {
+      if (isEmpty) "" else new String(bytes, charset)
 
-    def decodeString(charset: String): String =
+    override def decodeString(charset: Charset): String =
```
:+1:

ktoso replied:
so good to not be bound by JDK6 finally, isn't it ;)

viktorklang commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -124,29 +125,47 @@ object ByteString {
+      else toByteString1.take(n)
+
+    override def dropRight(n: Int): ByteString =
+      toByteString1.dropRight(n)
```
Not check `n <= 0` here?

ktoso replied:
right, might avoid some needless allocations - thanks!

viktorklang commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -124,29 +125,47 @@ object ByteString {
+      toByteString1.dropRight(n)
+
+    override def drop(n: Int): ByteString =
+      toByteString1.drop(n)
```
Not check n <= 0 here?

ktoso replied:
right, applied here too

viktorklang commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -164,11 +183,17 @@ object ByteString {
+      else new ByteString1(bytes, lowerBoundToZero(startIndex), lowerBoundToZero(length))
+
+    private final def lowerBoundToZero(i: Int): Int =
+      if (i <= 0) 0 else i
```
use `java.lang.Math.max(0, length)` since it is more likely to be intrinsified (i.e. remove lowerBoundtoZero)

ktoso replied:
right, good idea.

viktorklang commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -367,11 +428,81 @@ object ByteString {
+        else if (n <= bs.head.length) b.append(bs.head.take(n)).result
+        else take0(n - bs.head.length, b.append(bs.head), bs.tail)
+
+      take0(n, ByteString.newBuilder, bytestrings)
```
This allocates a new builder even if `n` <= 0 or `n` is as large, or larger, as `length`?

ktoso replied:
nice catch, will take care of it :)

ktoso commented:
Addressed the feedback by amending into the commit that introduced those optimisations

ktoso commented:
Impact on HTTP seems to be around +8%, for small ping/pong.

ktoso commented:
Some more measurements today:


```
Baseline was (initial, before this sprint):
  11 442 478 requests in 3.00m, 1.63GB read
Requests/sec:  63 571.07
Transfer/sec:      9.28MB

Previous result (copyToBuffer opt):
  19 631 499 requests in 3.00m, 2.80GB read
Requests/sec: 109 079.55
Transfer/sec:     15.92MB

  23 290 825 requests in 3.00m, 3.32GB read
Requests/sec: 129 413.42  (+18%)
Transfer/sec:     18.88MB (+18%)
```

viktorklang commented:
niiice

-- 
Cheers,
√

On Jul 20, 2016 11:16 AM, "Konrad Malawski" <notifications@github.com>
wrote:

> Some more measurements today:
>
> Baseline was (initial, before this sprint):
>   11 442 478 requests in 3.00m, 1.63GB read
> Requests/sec:  63 571.07
> Transfer/sec:      9.28MB
>
> Previous result (copyToBuffer opt):
>   19 631 499 requests in 3.00m, 2.80GB read
> Requests/sec: 109 079.55
> Transfer/sec:     15.92MB
>
>   23 290 825 requests in 3.00m, 3.32GB read
> Requests/sec: 129 413.42  (+18%)
> Transfer/sec:     18.88MB (+18%)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka/pull/20994#issuecomment-233896233>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAqdx4119rluPaLQ7vz99VFzAqcRzSKks5qXedSgaJpZM4JP0EG>
> .
>

ktoso commented:
Fixed MiMa issue (added method to ByteString).

drewhk commented on akka-actor/src/main/scala/akka/util/ByteIterator.scala:
```diff
@@ -249,7 +250,8 @@ object ByteIterator {
     }
 
-    @tailrec final override def drop(n: Int): this.type =
+    // For performance sensitive code, call drop() directly on ByteString (it's optimised there)
```
hm, shouldn't this be scaladoc comment?

ktoso replied:
I thought this is more of an "impl note for us", can move to scaladoc sure

drewhk commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -164,11 +185,14 @@ object ByteString {
     def apply(bytes: Array[Byte], startIndex: Int, length: Int): ByteString1 =
-      if (length == 0) empty else new ByteString1(bytes, startIndex, length)
+      if (length == 0) empty
+      else new ByteString1(bytes, Math.max(0, startIndex), Math.max(0, length))
```
shouldn't it throw if indices are out of bounds?

ktoso replied:
we're behaving like a `Seq` here, that's what it does:


```

scala> "abc".take(-1)
res0: String = ""

scala> "abc".drop(-1)
res1: String = abc

scala> "abc".drop(121212)
res2: String = ""

scala> "abc".take(121212)
res3: String = abc

```

drewhk commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -367,11 +427,83 @@ object ByteString {
+
+    override def dropRight(n: Int): ByteString = {
+      if (n <= 0) this
+      else if (n < bytestrings.last.length) {
```
cache bytestrings.last

ktoso replied:
nice catch, that's used a lot of times. thanks

drewhk commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -367,11 +427,83 @@ object ByteString {
+    override def take(n: Int): ByteString = {
+      @tailrec def take0(n: Int, b: ByteStringBuilder, bs: Vector[ByteString1]): ByteString =
+        if (bs.isEmpty || n <= 0) b.result
+        else if (n <= bs.head.length) b.append(bs.head.take(n)).result
```
cache bs.head

drewhk commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -367,11 +427,83 @@ object ByteString {
+      var fullDrops = 0
+      var remainingToDrop = n
+      do {
+        val currentLength = bytestrings(fullDrops).length
```
iterator instead of indexing?

drewhk commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -367,11 +427,83 @@ object ByteString {
+      do {
+        val currentLength = bytestrings(fullDrops).length
+        if (remainingToDrop >= currentLength) {
+          fullDrops += 1
```
maybe iterator instead?

drewhk commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -367,11 +427,83 @@ object ByteString {
+        if (remainingToDrop >= currentLength) {
+          fullDrops += 1
+          remainingToDrop -= currentLength
+        } else continue = false
```
just do a `remainingToDrop = 0` and remove the `continue` flag.

ktoso replied:
can't do that as `remainingToDrop > 0 && remainingToDrop < currentLength` is a value I need to keep.
Perhaps a crazy trick with making it `remainingToDrop = -remainingToDrop` would work hm, since then I still know I need to drop a bit

drewhk commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -367,11 +427,83 @@ object ByteString {
+        } else continue = false
+      } while (remainingToDrop > 0 && continue)
+
+      val remainingByteStrings = bytestrings.drop(fullDrops)
```
with an iterator you would be able to continue from here iterating the Vector

ktoso replied:
Good point, I'll try it

drewhk commented:
LGTM

patriknw commented:
Great work, @ktoso

ktoso commented:
Addressed all comments except the iterator rewrite – it can be done, but to gain performance from it we'd be forced to call a Scala `private[collection]` method (`.remainingVector`). I benchmarked that too (vs `.toVector`) and we'd get consistent 10m/s instead of fuzzy 8~10, it's not worth (yet) to call into the private APIs I thought.

Failure was unrelated, due to GC in param directives:

```
 [Times: user=11.05 sys=23.31, real=1.11 secs] 
5827.918: [GC cleanup 1693M->1693M(3072M), 0.0169310 secs]
 [Times: user=0.26 sys=0.03, real=0.02 secs] 
[info] - csv *** FAILED *** (1 second, 156 milliseconds)
[info]   Request was neither completed nor rejected within 1 second (ParameterDirectivesExamplesSpec.scala:203)
[info]   org.scalatest.exceptions.TestFailedException:
```

I'd like to merge so Johan's work shows improvements on top of this PR already

ktoso commented:
Merging.
